external/vcm/tests/test_limit.py

```diff
@@ -0,0 +1,99 @@
+import numpy as np
+import xarray as xr
+import pytest
+from vcm.limit import DatasetQuantileLimiter
+
+EPS = 1.0e-3
+
+
+def _xscaling(alpha):
+    scale = 1.0 / alpha
+    return np.arange(-scale, (scale + EPS))
+
+
+def _zscaling(alpha):
+    scale = 1.0 / alpha
+    return np.flip(np.arange(0.0, scale + EPS) / scale)
+
+
+def _get_arr(alpha):
+    xscaling = _xscaling(alpha)[:, np.newaxis]
+    zscaling = _zscaling(alpha)[np.newaxis, :]
+    return xscaling * zscaling
+
+    """        else:
+            upper = (scale - 0.5) * (np.arange(scale, 0.0 - EPS, -1.0) / scale)
+            lower = 0.5 * (np.arange(scale, 0.0 - EPS, -1.0) / scale)
+            expected_arr[-1, :] = upper
+            expected_arr[0 : (int(scale) + 1), :] = lower"""
+
+
+def get_dataset(alpha):
+    arr = _get_arr(alpha)
+    da = xr.DataArray(arr, dims=["x", "z"])
+    return xr.Dataset({"Q1": da, "Q2": da})
+
+
+def get_expected_arr(alpha, feature_dims, fit_indexers):
```

Comment:
This mocking code is pretty complex...two levels deep of if-statements and fairly complex random (?) data. I find this test harder to read then the code it is testing.

I think part of the issue is you are trying to unit-test the integration of the fitting and limiting, rather than each on their own. Something like this would be simpler IMO:

```
def test_fit_limits_correct():
    # ~ two lines for the data mock
    simple_data = ...
    limiter = DatasetQuantileLimiter.fit(simple_data)
    assert limiter._limits == ...


def test_transform():
    
    limiter = DatasetQuantileLimiter
    limiter._limits = simple mock
    assert limiter.transform(simple_data) == ...
```

```

Reply:
Those tests would be easier to refactor if we e.g. develop new ways of computing the limits.